File: js/packages/pyrmts/src/axis.ts
// Time-axis arithmetic. UTC throughout; calendar-correct for `mo`/`y` units
// (variable-width) and millisecond-multiplied for fixed-width units.

import type { Duration, Shard, TimeUnit } from './types.js'

const MS: Record<Exclude<TimeUnit, 'mo' | 'y'>, number> = {
  min: 60_000,
  h: 60 * 60_000,
  d: 24 * 60 * 60_000,
}

const SPAN_RE = /^(\d+)(min|h|d|mo|y)$/

export interface ParsedTimeSpan {
  count: number
  unit: TimeUnit
}

export function parseDuration(s: Duration | string): ParsedTimeSpan {
  const m = SPAN_RE.exec(s)
  if (!m) throw new Error(`Not a valid Duration: ${s}`)
  return { count: parseInt(m[1]!, 10), unit: m[2] as TimeUnit }
}

// Add `span` to a UTC instant. Calendar-aware for mo/y.
export function addSpan(t: Date, span: ParsedTimeSpan): Date {
  const { count, unit } = span
  if (unit === 'mo') {
    const r = new Date(t)
    r.setUTCMonth(r.getUTCMonth() + count)
    return r
  }
  if (unit === 'y') {
    const r = new Date(t)
    r.setUTCFullYear(r.getUTCFullYear() + count)
    return r
  }
  return new Date(t.getTime() + count * MS[unit])
}

// Floor a UTC instant to the start of its span. Supports count=1 for all
// units, and count>1 for fixed-width units (min/h/d) via ms division.
// Multi-unit calendar bins (e.g. `2mo`) aren't supported yet — alignment
// semantics aren't well-defined without an anchor.
export function floorToSpan(t: Date, span: ParsedTimeSpan): Date {
  const { count, unit } = span
  if (count !== 1) {
    if (unit === 'mo' || unit === 'y') {
      throw new Error(`Multi-unit calendar bins not supported: ${count}${unit}`)
    }
    const binMs = count * MS[unit]
    return new Date(Math.floor(t.getTime() / binMs) * binMs)
  }
  switch (unit) {
    case 'min':
      return new Date(Date.UTC(
        t.getUTCFullYear(), t.getUTCMonth(), t.getUTCDate(),
        t.getUTCHours(), t.getUTCMinutes(),
      ))
    case 'h':
      return new Date(Date.UTC(
        t.getUTCFullYear(), t.getUTCMonth(), t.getUTCDate(),
        t.getUTCHours(),
      ))
    case 'd':
      return new Date(Date.UTC(
        t.getUTCFullYear(), t.getUTCMonth(), t.getUTCDate(),
      ))
    case 'mo':
      return new Date(Date.UTC(t.getUTCFullYear(), t.getUTCMonth()))
    case 'y':
      return new Date(Date.UTC(t.getUTCFullYear(), 0))
  }
}

// Count bins of `bin` overlapping [from, to). Bins are span-aligned (floored
// at the start of each bin's boundary, not rolling from `from`).
export function binsInRange(from: Date, to: Date, bin: Duration): number {
  if (to <= from) return 0
  const span = parseDuration(bin)
  let cursor = floorToSpan(from, span)
  let n = 0
  while (cursor < to) {
    n++
    cursor = addSpan(cursor, span)
  }
  return n
}

// Enumerate shard periods covering [from, to]. Each period is the half-open
// interval [start, end) plus a `label` suitable for `{period}` substitution.
export function shardPeriodsCovering(
  from: Date,
  to: Date,
  shard: Shard,
): { start: Date; end: Date; label: string }[] {
  if (shard === 'all') {
    return [{ start: new Date(0), end: new Date(8.64e15), label: 'all' }]
  }
  if (shard === '1run') {
    throw new Error("'1run' shards are step-axis only; not yet supported")
  }
  const span = parseDuration(shard)
  const out: { start: Date; end: Date; label: string }[] = []
  let cursor = floorToSpan(from, span)
  while (cursor < to) {
    const next = addSpan(cursor, span)
    out.push({ start: cursor, end: next, label: formatPeriod(cursor, span) })
    cursor = next
  }
  return out
}

// Format a UTC instant as a `{period}` substitution string. The resolution
// of the label matches the shard's unit:
//   1y   → '2026'
//   1mo  → '2026-05'
//   1d   → '2026-05-24'
//   1h   → '2026-05-24T17'
//   1min → '2026-05-24T17-30'
export function formatPeriod(t: Date, span: ParsedTimeSpan): string {
  const yyyy = t.getUTCFullYear().toString().padStart(4, '0')
  const mm = (t.getUTCMonth() + 1).toString().padStart(2, '0')
  const dd = t.getUTCDate().toString().padStart(2, '0')
  const hh = t.getUTCHours().toString().padStart(2, '0')
  const mi = t.getUTCMinutes().toString().padStart(2, '0')
  switch (span.unit) {
    case 'y':   return yyyy
    case 'mo':  return `${yyyy}-${mm}`
    case 'd':   return `${yyyy}-${mm}-${dd}`
    case 'h':   return `${yyyy}-${mm}-${dd}T${hh}`
    case 'min': return `${yyyy}-${mm}-${dd}T${hh}-${mi}`
  }
}

File: python/pyrmts/src/pyrmts/axis.py
"""Time-axis arithmetic. UTC throughout; calendar-correct for `mo`/`y` units
(variable-width) and millisecond-multiplied for fixed-width units. Mirrors
`js/packages/pyrmts/src/axis.ts`."""
from __future__ import annotations

import re
from dataclasses import dataclass
from datetime import datetime, timezone

from .types import TimeUnit


_MS: dict[str, int] = {
    'min': 60_000,
    'h':   60 * 60_000,
    'd':   24 * 60 * 60_000,
}

_SPAN_RE = re.compile(r'^(\d+)(min|h|d|mo|y)$')


@dataclass(frozen=True)
class ParsedTimeSpan:
    count: int
    unit: TimeUnit


def parse_duration(s: str) -> ParsedTimeSpan:
    m = _SPAN_RE.match(s)
    if not m:
        raise ValueError(f"Not a valid Duration: {s!r}")
    return ParsedTimeSpan(count=int(m.group(1)), unit=m.group(2))  # type: ignore[arg-type]


def _ensure_utc(t: datetime) -> datetime:
    if t.tzinfo is None:
        return t.replace(tzinfo=timezone.utc)
    return t.astimezone(timezone.utc)


def add_span(t: datetime, span: ParsedTimeSpan) -> datetime:
    t = _ensure_utc(t)
    count, unit = span.count, span.unit
    if unit == 'mo':
        m = t.month - 1 + count
        y = t.year + m // 12
        return t.replace(year=y, month=m % 12 + 1)
    if unit == 'y':
        return t.replace(year=t.year + count)
    delta_ms = count * _MS[unit]
    return datetime.fromtimestamp((t.timestamp() * 1000 + delta_ms) / 1000, tz=timezone.utc)


def floor_to_span(t: datetime, span: ParsedTimeSpan) -> datetime:
    t = _ensure_utc(t)
    count, unit = span.count, span.unit
    if count != 1:
        if unit in ('mo', 'y'):
            raise ValueError(f"Multi-unit calendar bins not supported: {count}{unit}")
        bin_ms = count * _MS[unit]
        ms = int(t.timestamp() * 1000)
        floored = (ms // bin_ms) * bin_ms
        return datetime.fromtimestamp(floored / 1000, tz=timezone.utc)
    if unit == 'min':
        return t.replace(second=0, microsecond=0)
    if unit == 'h':
        return t.replace(minute=0, second=0, microsecond=0)
    if unit == 'd':
        return t.replace(hour=0, minute=0, second=0, microsecond=0)
    if unit == 'mo':
        return t.replace(day=1, hour=0, minute=0, second=0, microsecond=0)
    if unit == 'y':
        return t.replace(month=1, day=1, hour=0, minute=0, second=0, microsecond=0)
    raise AssertionError(f"unreachable: {unit}")


def bins_in_range(from_: datetime, to: datetime, bin: str) -> int:
    if to <= from_:
        return 0
    span = parse_duration(bin)
    cursor = floor_to_span(from_, span)
    n = 0
    while cursor < to:
        n += 1
        cursor = add_span(cursor, span)
    return n


@dataclass(frozen=True)
class ShardPeriod:
    start: datetime
    end: datetime
    label: str


def shard_periods_covering(from_: datetime, to: datetime, shard: str) -> list[ShardPeriod]:
    """Enumerate shard periods covering `[from, to)`."""
    if shard == '1run':
        raise NotImplementedError("'1run' shards are step-axis only; not yet supported")
    span = parse_duration(shard)
    out: list[ShardPeriod] = []
    cursor = floor_to_span(from_, span)
    while cursor < to:
        nxt = add_span(cursor, span)
        out.append(ShardPeriod(start=cursor, end=nxt, label=format_period(cursor, span)))
        cursor = nxt
    return out


def format_period(t: datetime, span: ParsedTimeSpan) -> str:
    """Format a UTC instant as a `{period}` substitution string."""
    t = _ensure_utc(t)
    yyyy = f"{t.year:04d}"
    mm = f"{t.month:02d}"
    dd = f"{t.day:02d}"
    hh = f"{t.hour:02d}"
    mi = f"{t.minute:02d}"
    unit = span.unit
    if unit == 'y':   return yyyy
    if unit == 'mo':  return f"{yyyy}-{mm}"
    if unit == 'd':   return f"{yyyy}-{mm}-{dd}"
    if unit == 'h':   return f"{yyyy}-{mm}-{dd}T{hh}"
    if unit == 'min': return f"{yyyy}-{mm}-{dd}T{hh}-{mi}"
    raise AssertionError(f"unreachable: {unit}")
